Ring docs (#1829)
ml-explore/mlx: docs/src/usage/launching_distributed.rst

:orphan:

.. _usage_launch_distributed:

Launching Distributed Programs
==============================

.. currentmodule:: mlx.core.distributed

Installing the MLX python package provides a helper script ``mlx.launch`` that
can be used to run python scripts distributed on several nodes. It allows
launching using either the MPI backend or the ring backend. See the
:doc:`distributed docs <distributed>` for the different backends.

Usage
-----

The minimal usage example of ``mlx.launch`` is simply

.. code:: shell

    mlx.launch --hosts ip1,ip2 my_script.py

or for testing on localhost

.. code:: shell

    mlx.launch -n 2 my_script.py

> The script connects to the provided host and launches the script on each host.
> It monitors each of the launched processes and terminates the rest if one of

Reviewer: The "script" here refers to two different things and is a little confusing:
```suggestion
The `mlx.launch` command connects to the provided host and launches the input script on each host.
It monitors each of the launched processes and terminates the rest if one of
```

them fails unexpectedly or if ``mlx.launch`` is terminated. It also takes care
of forwarding the output of each remote process to stdout and stderr
respectively.

Providing Hosts
^^^^^^^^^^^^^^^^

Hosts can be provided as command line arguments, like above, but the way that
allows to fully define a list of hosts is via a JSON hostfile. The hostfile has
a very simple schema. It is simply a list of objects that define each host via
a hostname to ssh to and a list of IPs to utilize for the communication.

.. code:: json

    [
        {"ssh": "hostname1", "ips": ["123.123.1.1", "123.123.2.1"]},
        {"ssh": "hostname2", "ips": ["123.123.1.2", "123.123.2.2"]}
    ]

Setting up Remote Hosts
^^^^^^^^^^^^^^^^^^^^^^^^

> In order to be able to launch the script on each host we need to be able to
> connect via ssh. Moreover the script and python executable need to be on each

Reviewer: ```suggestion
To use ``mlx.launch`` on each host we need to be able to
connect via ssh. Moreover ``mlx.launch`` and the input script need to be on each
```

Author: Man that's how badly I wrote this :-) ...

`mlx.launch` doesn't need to be on the hosts. But `python` does. So that meant the script and the actual python binary.

Reviewer: Oh of course.. I guess I also wasn't reading it that carefully sorry :). On second read "the script" is the user script and the "executable" is the `python` binary.

Reviewer: > I guess I also wasn't reading it that carefully sorry

Probably a good proxy for the average reader 😅

host and on the same path. A good checklist to debug errors is the following:

* ``ssh hostname`` works without asking for password or host confirmation
* the python executable is available on all hosts at the same path. You can use
  ``mlx.launch --print-python`` to see what that path is.

> * the script is available on all hosts at the same path

Reviewer: ```suggestion
* the `mlx.launch` script is available on all hosts at the same path
```
